codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketCloseFrameHandler.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.websocketx;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.AttributeKey;
+import io.netty.util.internal.ObjectUtil;
+
+/**
+ * Send {@link CloseWebSocketFrame} message on channel close, if close frame was not sent before.
+ */
+final class WebSocketCloseFrameHandler extends ChannelOutboundHandlerAdapter {
+    private static final AttributeKey<Boolean> CLOSE_FRAME_SENT_ATTR_KEY = AttributeKey
+        .valueOf(WebSocketCloseFrameHandler.class.getName() + ".CLOSE_FRAME_SENT");
+
+    private final WebSocketCloseStatus closeStatus;
+
+    WebSocketCloseFrameHandler(WebSocketCloseStatus closeStatus) {
+        this.closeStatus = ObjectUtil.checkNotNull(closeStatus, "closeStatus");
+    }
+
+    @Override
+    public void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        if (!closeFrameSent(ctx) && ctx.channel().isActive()) {
+            write(ctx, new CloseWebSocketFrame(closeStatus), ctx.voidPromise());
```

Comment:
imho you should call `writeAndFlush(...)` and attach a `ChannelFutureListener` that will close the channel after the write is done. That said you also will need to schedule some timeout that will close if the write did not complete in a period of time.

Reply:
while this idea looks good, it is not a correct place to implement. imho it should be either a separate/independent handler or an option in the ```FlushConsolidationHandler```. copy-pasting such logic from ```SslHandler``` looks incorrect.

my approach here works as following:

- for healthy connections - write + flush + close.
- for overflowed connections - write and flush could (in practice) be skipped, just close.

Reply:
anyway, if you insist on this change - I can add it

Reply:
I still think we should do something similar as in `SslHandler`. So yes please adjust it. Beside this using a `voidPromise` here could have some bad side-effects if the handlers in the pipeline to all correctly handle it.

Reply:
added

Reply:
I still would like to remove the `voidPromise()` here as there maybe other handles in the pipeline which not correctly handle the `unvoid()` thing. Please just use `ctx.newPromise()`

Reply:
changed as you wish.
however i don't understand... this void promise never leaves `WebSocketCloseFrameHandler`. how "other handlers" could be affected? could you please describe such scenario, so I could use this knowledge in the next PRs?